src/soc/intel/baytrail/txei.c

```diff
@@ -0,0 +1,1196 @@
+			printk(BIOS_ERR, "HECI: timed out reading answer!\n");
+			return 0;
+		}
+
+		if (!ipc_output_ready())
+			continue;
+
+		if (txe_filled_slots() >= cnt)
+			break;
+
+	}
+	return 1;
+}
+
+
+static size_t hdr_get_length(uint32_t hdr)
+{
+	return (hdr & TXEI_HDR_LENGTH) >> TXEI_HDR_LENGTH_START;
+}
+
+/* get number of full 4-byte slots */
+static size_t bytes_to_slots(size_t bytes)
+{
+	return ALIGN_UP(bytes, SLOT_SIZE) / SLOT_SIZE;
+}
+
+static int
+send_one_message(uint32_t hdr, const void *buff)
+{
+	size_t pend_len, pend_slots, remainder, i;
+	uint32_t tmp;
+	const uint32_t *p = buff;
+
+	if (!wait_heci_ipc_ready())
+		return 0;
+
+	/* Get space for the header */
+	if (!wait_write_slots(1))
+		return 0;
+
+	/* First, write header */
+	write_slot(0, hdr);
+
+	pend_len = hdr_get_length(hdr);
+	pend_slots = bytes_to_slots(pend_len);
+
+	if (!wait_write_slots(pend_slots))
+		return 0;
+
+	/* Write the body in whole slots */
+	i = 0;
+	while (i < ALIGN_DOWN(pend_len, SLOT_SIZE)) {
+		write_slot(i + 4, *p++);
+		i += SLOT_SIZE;
+	}
+
+	remainder = pend_len % SLOT_SIZE;
+	/* Pad to 4 bytes not touching caller's buffer */
+	if (remainder) {
+		memcpy(&tmp, p, remainder);
+		write_slot(i + 4, tmp);
+	}
+
+	set_ipc_input_doorbell();
+	host_gen_interrupt();
+
+	/* Make sure nothing bad happened during transmission */
+	if (!txe_ready())
+		return 0;
+
+	return pend_len;
+}
+
+static bool heci_can_send(void)
+{
+	pci_devfn_t dev = PCI_DEV(0, TXE_DEV, TXE_FUNC);
+	uint32_t fw_sts0;
+
+	if (!is_txe_devfn_visible())
+		return false;
+
+	if (txe_is_excluded_from_bios_flows()) {
+		printk(BIOS_INFO, "TXE excluded from platform flows\n");
+		printk(BIOS_ERR, "Can not send HECI messages\n");
+		return false;
+	}
+
+	if (txe_has_error()) {
+		printk(BIOS_ERR, "TXE Error occurred, not sending HECI message\n");
+		return false;
+	}
+
+	if (txe_fw_update_in_progress()) {
+		printk(BIOS_INFO, "TXE FW is updating, not sending HECI message\n");
+		return false;
+	}
+
+	fw_sts0 = pci_read_config32(dev, TXE_FWSTS0);
+
+	switch ((fw_sts0 & TXE_CURR_OPMODE_MASK) >> TXE_CURR_OPMODE_SHIFT) {
+	case TXE_FWSTS0_COM_NORMAL:
+	case TXE_FWSTS0_COM_SECOVER_TXEI_MSG:
+		break;
+	case TXE_FWSTS0_COM_DEBUG:
+	case TXE_FWSTS0_COM_SOFT_TEMP_DISABLE:
+	case TXE_FWSTS0_COM_SECOVER_JMPR:
+		printk(BIOS_INFO, "Current TXE operation mode does nto allow sending HECI messages\n");
+		return false;
+	}
+
+	return true;
+}
+
+static enum txe_tx_rx_status
+heci_send(const void *msg, size_t len, uint8_t host_addr, uint8_t client_addr)
+{
+	uint8_t retry;
+	uint32_t hdr;
+	size_t sent, remaining, cb_size, max_length;
+	const uint8_t *p;
+
+	if (!msg || !len)
+		return TXE_TX_ERR_INPUT;
+
+	if (!heci_can_send())
+		return TXE_TX_ERR_INPUT;
+
+	for (retry = 0; retry < HECI_MAX_RETRY; retry++) {
+		p = msg;
+
+		if (check_heci_reset())
+			txe_txei_init();
+
+		if (!wait_heci_ipc_ready()) {
+			printk(BIOS_ERR, "HECI: not ready\n");
+			continue;
+		}
+
+		cb_size = 16;
```

Comment:
What is it? Why isn't this a constant? In fact, why `max_length` isn't a constant?

Reply:
https://github.com/Dasharo/coreboot/pull/542/files#diff-9c3c7de7fe3776cfe7cd7ceaf5eaa33f7f23d275b737008f284189226f44ed2bR660
https://github.com/Dasharo/coreboot/pull/542/files#diff-9c3c7de7fe3776cfe7cd7ceaf5eaa33f7f23d275b737008f284189226f44ed2bR20-R21